File: 016/main.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var threeSumClosest = function (nums, target) {
  let len = nums.length
  if (len === 3) {
    return nums.reduce((total, num) => {
      return total + num
    }, 0)
  }
  nums.sort((a, b) => {
    return a > b ? 1 : -1
  })
  let res = nums[0] + nums[1] + nums[2]
  let min = Math.abs(res - target)
  for (let i = 0; i < len; i++) {
    if (i > 0 && nums[i] === nums[i - 1]) continue
    let start = i + 1
    let end = len - 1
    let sum
    while (start < end) {
      sum = nums[i] + nums[start] + nums[end]
      if (Math.abs(sum - target) <= min) {
        min = Math.abs(sum - target)
        res = sum
        if (sum < target) start++
        else end--
      } else if (Math.abs(sum - target) > min && sum > res) {
        end--
      } else if (Math.abs(sum - target) > min && sum < res) {
        start++
      }
    }
  }
  return res
}

console.log(threeSumClosest([-1, 2, 1, -4], 1))
console.log(threeSumClosest([0, 0, 0], 1))

File: 016/main.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <math.h>
using namespace std;

class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        int len = nums.size();
        sort(nums.begin(), nums.end());
        int res = nums[0] + nums[1] + nums[2];
        int min = abs(res - target);
        if (len == 3) return res;
        for (int i = 0; i < len; i++) {
            if (i > 0 && nums[i] == nums[i - 1]) continue;
            int start = i + 1;
            int end = len - 1;
            while (start < end) {
                int sum = nums[i] + nums[start] + nums[end];
                if (abs(sum - target) <= min) {
                    min = abs(sum - target);
                    res = sum;
                    if (sum < target) start++;
                    else end--;
                } else if (sum > res) end--;
                else if (sum < res) start++;
            }
        }
        return res;
    }
};

int main()
{
    Solution sol;
    vector<int> nums;
    nums.push_back(-1);
    nums.push_back(2);
    nums.push_back(1);
    nums.push_back(4);
    int res = sol.threeSumClosest(nums, 1);
    cout << res << endl;
    return 0;
}
